Ctrl+Click to add a model, Alt+Click to reset before add

eyal0 opened the pull request:
* Ctrl+Click will add a model to the platform without clearing and without putting the modal dialog.
* Alt+Click will clear the platform and then add the model without the modal dialog.
* With neither Alt nor Ctrl, the modal dialog shows like usual with a Hint there explaining how to use Ctrl and Alt.

This makes is much faster to load multiple models.

I tested ctrl and alt for:
* magic wand button
* upload button
* drag-n-drop upload

All of them work with ctrl, alt, and without.

This fixes https://github.com/eyal0/OctoPrint-Slicer/issues/3

kennethjiang commented:
I made a comment a few hours back, but somehow it's missing from here. Pardon me if you receive 2 comments with similar content.

I love the way of using modifier keys to skip this popup modal, which I have been wanting to get rid of because it can be annoying.

However, I ran into a few issues with your PR:
1. `Ctrl` key doesn't work in Chrome or Safari on Mac. It does work in Firefox. This can be solved simply by using `Shift` (unless `shift` key has issue on Windows).
1. When user drag-and-drop multiple STL files into OctoPrint, I ran into quite a few issues depending on which key I was holding.
1. When user upload or drag-n-drop STL files, `self.slicingViewModel.show` is not called when the user release mouse button. Instead it is called when STL is done uploading, at which point user may have already released `Alt/Ctrl/Shift` key. This is especially true for large STL files. Try a 100MB STL file you will definitely see the problem.

I have been working on this for a few hours now. *#2* and *#3* turn out to be very tricky to solve perfectly without introducing new edge cases. 

Here is one way that I hope we can implement with reasonable robustness: We only accept 1 modifier key to indicate *clear the bed and start a new session*. When users want to add STL to existing files, they still have to click "OK" on the pop-up modal every time.

Let me know if you have other ways that can be implemented with reasonable robustness. Otherwise I can take a swing at doing what I proposed above.

eyal0 commented:
I didn't use shift because when I tried shift and click on the magic wand for two different STL files, all the text between got highlighted. Ctrl and alt didn't do that. I guess you could make shift act the same as Ctrl.  I don't have a Mac so I can't test it.

For number 3, we need to track when the button is actually released. We could plug in to the dragndrop event handling in octoprint and try to capture it when it happens.

Multiple file upload is poorly supported in octoprint anyway. I wouldn't​ expect it to work all that well. I guess that I'll keep working on it...

eyal0 commented:
Okay, I just tested by slowing down the network in Chrome.  There are three ways to upload:
1. Magic wand button
2. Upload button.
3. drag n drop.

For me, magic wand button and upload button work if I press and release ctrl while I click with the mouse button.  It works even if I release the ctrl button way before the upload finishes.

For DragNDrop, I see that it doesn't work.  I had the ctrl key held down while I released the mouse button and the ctrl key didn't have effect.

So for me, the only problem is the drag and drop.  Is that what you're seeing, too?

BTW, I don't like the idea of having just one modifier key because usually when I want to add more than one STL, I want to add lots of them.  For example, I want to print 10 little soldiers so I'd have 10 dialog boxes.

I think that we should try to get ctrl (or shift) and alt to work.  And in the future, the dialog box could maybe have a numeric input so that users that want to add 10 all at once can do that.  I think that there is an issue open for you on that.

kennethjiang commented:
yes drag-n-drop is where the modifier keys (for both alt and ctrl/shift) are broken. Ctrl is broken in Chome and Safari (on Mac) but it's a separate and easier issue.

True that having to click through the modal multiple times is annoying. But I'll implement the function to duplicate parts. Assuming that duplicate function is implemented, do you think that having only 1 modifier key will be easier to implement correctly, as well as offer a simpler user interaction?

eyal0 commented:
I think that I still prefer two modifiers.  It feels weird having a modifier for just one option and not the other.

On the Mac, maybe it makes more sense to use Cmd instead of Ctrl?  On the PC, ctrl-c ctrl-v are copy/paste but on the Mac, it's cmd c and v.  In eclipse, cmd+space is used on mac but ctrl+space on PC.  So I think that cmd might be a good choice.  I'm not sure.  I need to find a Mac on which to test.  I'd really like to have two modifiers to match the two buttons in the dialog.  Worst case, allow shift in addition to ctrl for adding without clearing.  That should work on Mac.

I updated the branch to handle the drag and drop bug.  Let me know if it works for you.

eyal0 commented:
Okay, I made a version that should work for drag and drop and upload button and magic wand.  I simulated a slow upload time to make sure that I'm releasing the modifier before the upload completes and it seems okay.

I also added support for the meta key, which according to what I read online is the Cmd key on a Mac.  So Cmd+click should work for adding to the platform.  Like I wrote before, some things that are Ctrl+... on PC are Cmd+... on the Mac so this seems like a good choice.

Finally, I customize the modal depending on mac or not to show "Ctrl" or "Cmd" in the hint.

Please test it and let me know.  Thanks!

kennethjiang commented:
Will look at your new commit when I get a chance.

Have you tested the case of drag-and-drop'ing multiple STL files at a time?

kennethjiang commented:
I've found it a common practice to use 1 modifier to "skip to" most-often option, rather than to assign shortcuts to each and every options. Why do you think it's weird?

I have tried very hard to keep things as simple as they can be, in both user interactions as well as code. I assume that after "duplicate model" is implemented, adding new STL to existing ones will be minority of the use cases, therefore users won't be too annoyed (they may even appreciate the explicitly) by the popup modal. Can you think of specific use cases that this popup modal will become too annoying after "duplicate model" is implemented?

kennethjiang commented:
I like your way of overriding `fileViewModel` to capture when the file starts/ends uploading. It works better than my old code when it comes to multiple STL files!

However `Cmd` works on Chrome but not Safari. I guess I will merge your PR with only `Alt` modifier key unless you feel strongly why this is a bad idea.

eyal0 commented:
What if I have multiple files that I want to put on the platform:

![image](https://cloud.githubusercontent.com/assets/109809/25369967/39c08326-2990-11e7-8a5d-a45c09f85a11.png)

I want one of each of those.  I will need to click the magic wand and then the modal, back and forth, many times.  I think that this is a good use case for having Ctrl.  If Ctrl worked, I could hold it down and add all those files easily.

If Cmd or Ctrl are having problems on different platforms, we can work on that.  I think that the ctrl ability is too useful to give up.  At worse, we could allow shift to work like ctrl.  That should work, even on a mac.

I use Linux and it's working fine on all computers.

eyal0 commented:
Maybe this is the same issue?  http://stackoverflow.com/a/20024823/4454

I'll borrow a mac and try monitoring the keyboard.  In the meanwhile, you can type this in the dev console:

`monitorEvents(document, ["keypress", "keyup", "keydown"])`

and it'll show you all the keypresses.  You can see what code is used for Cmd.

Finally, maybe we can use this library as a platform-agnostic way to get cmd/ctrl/alt, etc: https://github.com/madrobby/keymaster .

I really like ctrl.  If you get read of it, I'll probably just keep it in my branch, which is okay, too.

kennethjiang commented:
I tested it on Firefox `Cmd` key works. The only problem on Mac now is Safari.

I'll appreciate if you can test all main stream browsers on Windows (including IE 11+). We still leave out Linux but I don't think we have time to do that for now.

I agree there are situations when you will want to *add* not *new*. But my experience is <10% of the times I'd want to *add* multiple models to the bed. And when I did, I'd just drag-and-drop all of them from file system, in which case I'm getting only 1 pop-up, which doesn't bother me (actually in that case I prefer a pop-up modal to make it explicit to me what will happen). But you and other users may have different situations. Can you further elaborate why you will need `Ctrl` key *frequently*? 

PS. Let me explain a bit why I care so much about 1 modifier key. We wrote the code so we know exactly how to use the software, no matter how complicated it is. But we need to consider thousands of users who are actively using this plugin but barely know what "Cura profile" is (I just received an email about this from a user). My hunch is even if we simplified it to 1 modifier key, only the 10% advanced users will eventually figure out how to use it effectively.

eyal0 commented:
The use case for me is like the image above. Lots of magic wand buttons to press in a row. If I can hold down Ctrl while I press them, that solves the problem for me.

All my testing has been on Linux, not Windows. I can confirm that it works, Ctrl and alt with chrome. I didn't check Firefox.

If you really want just want modifier, we can change this PR to do just one modifier and I'll have two modifiers on my own fork. But I don't see the harm in two modifiers. It offers a feature without slowing the interface.

kennethjiang commented:
I did some more testing...

* `Cmd` doesn't work in Chrome. I remember testing it on Chrome last night and it worked. But today I tried many times and it doesn't work. I made some mistakes yesterday.
* In Firefox, when I tried to use `Cmd` and `Alt` multiple times, alternating between them, more than 50% of the times the modal would pop up. I tested with drag-n-drop'ing 3 STL files at a time. Can you repeat similar testing to see if you are getting the same issue?

eyal0 commented:
I got an opportunity to try it on Chrome on a Mac laptop.  It didn't work.  As far as I can tell, it didn't even try to upload!  I have some ideas for solutions that I'm working on.

eyal0 commented:
I think that I've fixed Cmd.  For good measure, I added shift, too.  It was very complicated because the behavior of dropEffect and effectAllowed on drag and drop on mac vs linux are different.  But now it should work pretty well.  Still having problems on firefox, though.  Grr.

kennethjiang commented:
I'm concerned by the increased complexity of this PR. The old code around loading STLs works flawlessly in all tests I have done myself. But I'm already receiving 2-5 errors per day. And they are extremely difficult to triage and debug.

This PR doesn't seem to be in a direction that will help with the situation. And since I highly doubt that most users can figure out how to use `Alt/Cmd/Ctrl/Shift` effectively, I started to think the added function won't be worth the increased complexity of the code and risk of bugs.

I'd like to propose a much simpler solution: add a checkbox to the modal that says "remember my decision until the browser window/tab is closed". If user checks that checkbox we save a boolean to session storage. Every time before we pop up modal we check for that boolean value in session storage. It achieves similar goals as shortcut keys. The code will be much simpler. And more importantly, more users will be able understand how to use it effectively.

Let me know what you think.

John-Mc commented:
I've been trying to follow this as best I can. From the point of view of a user (one who uses a Mac at home, and Windows or a Chromebook when assisting at an elementary school), I like the idea of a checkbox or some other method that effectively says "keep doing this until I tell you otherwise". 

How to make the "until I tell you otherwise" part clear is another question. Would it complicate things too much or mess up the interface to have something show up on the screen to inform users that they are in "multiple file add mode"  maybe with something that also says: "do XYZ to cancel/exit" (click another tab or hit ctrl-X ??). Not sure if it makes sense for other functions in the slicer tab to be greyed out until they exit this mode?

eyal0 commented:
My goal with the change was to make it faster to add models.  I like the solution that I made because, for new users, nothing is changed but they get a hint about how to work faster.  If they know about the hint, they get to work faster.  And using ctrl+drag or alt+drag seems natural because that's already common in Windows and I think Mac, too.

Here's a speed comparison of how many clicks per scenario.  w is wand, m is modal, c is checkbox.

|          | Print 3 models at once                           | 3 models in a row                                | 2 models then 2 models      |   |
|----------|--------------------------------------------------|--------------------------------------------------|-----------------------------|---|
| Current  | wand, modal, wand, modal, wand, modal (**6 clicks**) | wand, modal, wand, modal, wand, modal (**6 clicks**) | wmwmwmwm (**8 clicks**)         |   |
| Ctrl/Alt | holding ctrl: wand, wand, wand (**3 clicks**)        | holding shift: wand, wand, wand (**3 clicks**)       | wwwwww (**4 clicks**)           |   |
| checkbox  | wand, checkbox, wand, wand (**4 clicks**)            | wcww, (**4 clicks**)                                 | wcw, reload, wcw (**7 clicks**) |   |

Using ctrl/alt is the fastest in all scenarios.

@John-Mc  If you want to try it out, install this:  https://github.com/eyal0/OctoPrint-Slicer/archive/new_master.zip .  I also have collision detection there, another arranging algorithm, and a different lay flat algorithm.

John-Mc commented:
Thanks @eyal0. I don't have a printer at home, but I do have a RPi3, so I can poke around it when I get a chance.

kennethjiang commented:
@John-Mc Good point on having a visual cue to hint users which "mode" they are currently in. Whereas greying out the whole slicer tab is relatively easy, doing so for "other functions but the slicer window" is technically challenging or even impossible.

I will ask a friend of mine who is a UI designer (who designed the buttons for me) for his opinion on what kind of visual cue is appropriate, and assess how easily that can be implemented. Before that is done, do you think it's a reasonable middle ground to say "remember my choice until I close this browser window/tab"?

kennethjiang commented:
I consulted my design friend. His opinion is to make user interaction simpler and less cluttered. Based on his suggestions, I opened #100 

@eyal0 @John-Mc please comment on #100 if you have different opinions.
